File: languages/cpp/projects/04-expert/main.cpp
#include <chrono>
#include <iostream>
#include <memory>
#include <string>
#include <vector>

class Step {
public:
    explicit Step(const std::string& nameValue) : name(nameValue), processedCount(0) {}

    void run() {
        ++processedCount;
    }

    const std::string& getName() const {
        return name;
    }

    int getProcessedCount() const {
        return processedCount;
    }

private:
    std::string name;
    int processedCount;
};

int main() {
    const int jobCount = 3;

    std::vector<std::unique_ptr<Step>> steps;
    steps.push_back(std::unique_ptr<Step>(new Step("load")));
    steps.push_back(std::unique_ptr<Step>(new Step("transform")));

    const auto start = std::chrono::high_resolution_clock::now();

    for (int job = 0; job < jobCount; ++job) {
        (void)job;
        for (auto& step : steps) {
            step->run();
        }
    }

    const auto end = std::chrono::high_resolution_clock::now();
    const auto elapsed = std::chrono::duration_cast<std::chrono::microseconds>(end - start).count();

    std::cout << "Running " << jobCount << " jobs through " << steps.size() << " steps...\n";
    for (const auto& step : steps) {
        std::cout << "Step " << step->getName() << " processed " << step->getProcessedCount() << " jobs\n";
    }
    std::cout << "Elapsed (microseconds): " << elapsed << '\n';

    return 0;
}

File: languages/csharp/projects/04-expert/main.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;

sealed class Step
{
    private readonly string name;
    private int processedCount;

    public Step(string nameValue)
    {
        name = nameValue;
        processedCount = 0;
    }

    public string Name => name;

    public int ProcessedCount => processedCount;

    public void Run()
    {
        processedCount++;
    }
}

class Program
{
    static void Main()
    {
        const int jobCount = 3;

        List<Step> steps = new List<Step> { new Step("load"), new Step("transform") };

        Stopwatch stopwatch = Stopwatch.StartNew();

        for (int job = 0; job < jobCount; job++)
        {
            foreach (Step step in steps)
            {
                step.Run();
            }
        }

        stopwatch.Stop();
        long elapsedMicroseconds = (long)(stopwatch.Elapsed.TotalMilliseconds * 1000.0);

        Console.WriteLine($"Running {jobCount} jobs through {steps.Count} steps...");
        foreach (Step step in steps)
        {
            Console.WriteLine($"Step {step.Name} processed {step.ProcessedCount} jobs");
        }

        Console.WriteLine($"Elapsed (microseconds): {elapsedMicroseconds}");
    }
}
